### DISCUSSION TITLE
Fill-in code

### DISCUSSION THREAD

**Discussion by jonatanschroeder** [PrairieLearn/PrairieLearn#6037] [Q&A]
Has anyone ever added a `pl-integer-input` (or other similar inputs) inside a `pl-code`? `pl-overlay` seems to work but is a bit clumsy:

<img width="316" alt="image" src="https://user-images.githubusercontent.com/1004907/180897569-63560bc4-51b8-4d83-8f0e-2869b860e050.png">

**Comment by echuber2** [comment]
Another clumsy way to do it would be to use a variable for the unknown value and have the input prompt after the code segment. (Looks like I can't mark this comment as anything but "Answer" despite it being a non-answer.)

**Comment by glherman** [comment]
You can add a bit css to make the input box smaller/tighter so that it's less clumsy. We've done a bit of stuff like this in my class and have a modification of the pl-string element we call pl-embed that we use for everything from annotating diagrams and code to having students fill in tables and other diagrams.

**Reply by nealterrell** [reply]
Are you able to share that element?

**Reply by glherman** [reply]
sorry it's not well documented. https://github.com/PrairieLearn/pl-cs233/tree/master/elements/pl-string-embed. There are probably a few things that need to be improved stylistically, but I haven't gotten around to them because I haven't shared this before

**Reply by glherman** [reply]
Example uses:
https://github.com/PrairieLearn/pl-cs233/tree/master/questions/datapath_MIPS_data
https://github.com/PrairieLearn/pl-cs233/tree/master/questions/pipeline_perf_py

**Reply by nealterrell** [reply]
I appreciate the response! But I don't have access to that repository, possibly because I'm not at UIUC.

**Reply by glherman** [reply]
just gave you read access

**Comment by nealterrell** [comment]
@jonatanschroeder and @glherman -- we're experimenting with pl-overlay like you. We've found that if  a student changes the font size of their browser -- using a browser setting, not using Zoom In (which does work) -- the overlaid elements become mispositioned. Any thoughts on how to address that?

**Reply by glherman** [reply]
We used these types of questions throughout the pandemic with no CBTF. We've run >1000 students through exams with these questions and have never heard a single complaint. So, we haven't tried to address the issue. It doesn't seem to be a problem...yet.

**Reply by glherman** [reply]
My students are not shy about complaining about EVERY formatting or interface issue that they don't like, so I think it is legitimately a non-issue for the students.

**Reply by jonatanschroeder** [reply]
I used pl-overlay in the exam, but it was in a controlled lab (a lite version of CBTF), so this was not an issue. I haven't tried it in a quiz setting yet, though.

**Reply by echuber2** [reply]
@nealterrell Are the students seeing the breakage using any particular browser/platform? I played around with a previous version of 233's element and I thought that Safari did unusual things with the form widths. Even if that's the case, if a browser is messing with the form widths based on a base font size, it might be fixable with some additional CSS.

**Reply by nealterrell** [reply]
@echuber2 I haven't deployed this question yet, just done some very light experimentation. Chrome will misposition the text boxes if Settings are used to change the default Font Size. (Makes sense; the overlay coordinates are pixel coordinates, yes? And we are measuring those using the default rendering of the page, which changing the Font Size disrupts.)

**Reply by echuber2** [reply]
I thought the px unit wasn't supposed to depend on the base font size, but rather the display DPI. (It's not an actual pixel unit, but a magic unit the CSS (or rather, the W3C) defines specially.) What OS?

**Reply by echuber2** [reply]
Now it occurs to me that some browsers may actually be using the base font size setting _in order to_ detect the device DPI

**Reply by nealterrell** [reply]
@echuber2 here is a before and after comparison.

![image](https://user-images.githubusercontent.com/10083341/181845328-abe6260d-2a17-4340-b53d-6331b573a2e6.png)

![image](https://user-images.githubusercontent.com/10083341/181845344-256757f0-92e6-4cd5-b022-31cd3ec6375a.png)

**Reply by nealterrell** [reply]
This is on Chrome, macOS.

**Reply by glherman** [reply]
So, the issue is that the size of the overlay background is changing based on the font size but the pixel locations of the embedded input boxes are staying the same. I think the only way around this is to somehow read the browser state and dynamically adjust your input box locations accordingly. If I knew how to do that...
